src/web/js/output-ui.js

```diff
@@ -134,278 +218,55 @@
+          viewportMax = scrollInfo.clientHeight + viewportMin;
+        } else {
+          var repl = document.querySelector('.repl');
+          viewportMin = repl.scrollTop;
+          viewportMax = viewportMin + repl.scrollHeight;
         }
-      });
-
-      elt.on("mouseleave", function() {
-        $("."+locKey).removeClass("hover");
-        warnDesired = 0;
-        setTimeout(function() { setWarningState(jQuery(".warning-upper"));},
-                   warnWait);
-        setTimeout(function() { setWarningState(jQuery(".warning-lower"));},
-                   warnWait);
-      });
-
-      elt.on("click", function() {
-        warnDesired = 0;
-        jQuery(".warning-upper").fadeOut("fast");
-        jQuery(".warning-lower").fadeOut("fast");
-        editor.scrollIntoView(cmLoc.start, 100)
-      });
-    }
-
-    function hoverLocs(editors, runtime, srcloc, elt, locs, cls) {
-      var get = runtime.getField;
-      var cases = runtime.ffi.cases;
-
-      function highlightSrcloc(s, cls, withMarker) {
-        return runtime.safeCall(function() {
-          return cases(get(srcloc, "is-Srcloc"), "Srcloc", s, {
-            "builtin": function(_) { /* no-op */ },
-            "srcloc": function(source, startL, startC, startCh, endL, endC, endCh) {
-              var cmLoc = cmPosFromSrcloc(runtime, srcloc, s);
-              var editor = editors[source];
-              if(editor) {
-                return editor.markText(
-                  cmLoc.start,
-                  cmLoc.end,
-                  { className: cls });
-              } else {
-                return null;
-              }
-            }
-          });
-        }, withMarker, "highlightSrcloc");
-      }
-
-      // There are warnings to show indicating whether the check output
-      // is from a check that is currently off screen.  The problem is
-      // that the repl output is a bunch of little elements, so moving
-      // from one line to another of the check output is likely to
-      // trigger a mouseleave and then a mouseenter event. Showing the
-      // warnings in a sensible way (not having them flicker in and out)
-      // involves ignoring some of these transitions.
-      //
-      // The other problem is that the mouse events are not 100%
-      // reliable. That is, a fast swipe through a bunch of small areas
-      // is probably not going to provide an even number of enter and
-      // leave events. More likely one or more enter and leave events
-      // will be skipped.  So we arrange for a fadeout of the warning
-      // after a few seconds.
 
-      // Set this to the opacity desired for the warning.  This should
-      // be either fadeAmt or zero.
-      var warnDesired = 0;
-      // These are just fixed parameters and can be eliminated if
-      // efficiency demands it.
-      var warnWait = 250;
-      var warnDuration = 5000;
-      var fadeAmt = 0.5;
+        var direction;
+        var TOP     = 0,
+            BOTTOM  = 1;
```

Comment:
I usually do one `var` statement per, because it makes moving code around later easier.  But I don't have strong feelings about it.